developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+doesn't seem to be a way on the host to uniquely and securely identify guests.
+If there is indeed no such way, our only possibility seems to be using
+cryptography for that.
+
+Guest topology files contain virtual DAI components of type "dai_out" and
+"dai_in." Their data tuples contain a new token of type SOF_TKN_COMP_REF, which
+contains an ID of the mixer or the demux, to which this pipeline should be
+connected.
+
+After obtaining the topology, the guest SOF instance uses a second new
+SOF_IPC_TPLG_VFE_COMP_ID VirtIO-specific IPC message to request its allocated
+component ID base. The host maintains a bitmap of guest component ID ranges to
+guarantee, that component IDs stay unique regardless of the order in which
+guests are brought up and down.
+
+Next the guest proceeds by sending the parsed topology over IPC to the host,
+which is then forwarded to the DSP firmware. The guest uses the .send_msg()
+callback to forward any IPC, including topology related, with no processing at
+all, to the host. The only exception is blocking IPCs when the guest is
+resuming, while the host doesn't need to be resumed (see 3.3 section on "reset
+count" for details).
+
+The host receives the guest topology over the IPC virtual queue. In most cases
+the host IPC handler just forwards IPCs to the DSP, but in some cases it has to
+handle or modify IPC contents locally. One such example is linking the guest and
+host topology fragments together. As mentioned above, guest topology files
+contain a new token, referencing mixer or demux components in the host topology.
+This value is then assigned to the .ref_comp_id field of struct
+sof_ipc_comp_config during topology file parsing and sent by the guest to the
+host. The host then uses that value to modify connection IPC messages from the
+guest before sending them to the DSP.
+
+3.2. Streaming
+--------------
+
+Guest audio streaming is mostly transparent for the host. The host audio
+subsystem doesn't get involved with most guest streaming or kcontrol operations.
+All those operations are completely processed on the guest and only resulting
+DSP IPCs are forwarded to the host. The only exceptions are beginning and ending
+of the streaming, at which times the ASoC subsystem on the host has to be
+informed, that the respective virtual pipeline, typically including a physical
+codec DAI, is becoming active or inactive respectively. Deactivation of the
+guest pipeline on the host requires changes to the ASoC core, without which the
+guest audio path is not recognised as inactive and is not pruned, thus making
+further streaming by the guest impossible.
+
+Upon reception of the STREAM_TRIG_START IPC message the vhost driver activates
+the virtual PCM pipeline and updates the routing information on the host.
+
+As mentioned above, it is our goal to implement data zero-copy. As long as
+this isn't the case, the guest VirtIO driver implements .block_write() and
```

Comment:
You should really have two models
- copy over virtio queues
- zero-copy

and a means to switch between the two for debug.

Reply:
We don't have 0-copy yet, so I'm trying not to speak too much about it. I only mention it to explain our eventual intention and to indicate where we already expect differences. But yes, once we get 0-copy, we might also keep the current implementation as a fall-back.

Reply:
what I meant is that we don't necessarily want to use ONLY zero-copy in the end, it's useful to keep the mechanism with copies to e.g. debug issues, proceed with faster integration on specific platforms, etc.

You present the solution as exclusive, and I believe keeping the copy-based solution can be useful at times.